semantic-core/src/Data/Core.hs

```diff
@@ -83,69 +96,88 @@ deriving instance (Ord  a, forall a . Eq   a => Eq   (f a)
-  Unit      >>=* _ = Unit
-  Bool b    >>=* _ = Bool b
-  If c t e  >>=* f = If (c >>= f) (t >>= f) (e >>= f)
-  String s  >>=* _ = String s
-  Load b    >>=* f = Load (b >>= f)
-  Edge e b  >>=* f = Edge e (b >>= f)
-  Frame     >>=* _ = Frame
-  (a :. b)  >>=* f = (a >>= f) :. (b >>= f)
-  (a := b)  >>=* f = (a >>= f) := (b >>= f)
-  Ann l b   >>=* f = Ann l (b >>= f)
+  Rec b      >>=* f = Rec ((>>=* f) <$> b)
+  (a :>> b)  >>=* f = (a >>= f) :>> (b >>= f)
+  (a :>>= b) >>=* f = ((>>= f) <$> a) :>>= (b >>=* f)
+  Lam b      >>=* f = Lam ((>>=* f) <$> b)
+  (a :$ b)   >>=* f = (a >>= f) :$ (b >>= f)
+  Unit       >>=* _ = Unit
+  Bool b     >>=* _ = Bool b
+  If c t e   >>=* f = If (c >>= f) (t >>= f) (e >>= f)
+  String s   >>=* _ = String s
+  Load b     >>=* f = Load (b >>= f)
+  Record fs  >>=* f = Record (map (fmap (>>= f)) fs)
+  (a :. b)   >>=* f = (a >>= f) :. b
+  (a := b)   >>=* f = (a >>= f) := (b >>= f)
+  Ann l b    >>=* f = Ann l (b >>= f)
+
+
+rec :: (Eq a, Carrier sig m, Member Core sig) => Named a -> m a -> m a
+rec (Named u n) b = send (Rec (Named u (abstract1 n b)))
+
+(>>>) :: (Carrier sig m, Member Core sig) => m a -> m a -> m a
+a >>> b = send (a :>> b)
+
+infixr 1 >>>
 
+unseq :: (Alternative m, Member Core sig) => Term sig a -> m (Term sig a, Term sig a)
+unseq (Term sig) | Just (a :>> b) <- prj sig = pure (a, b)
+unseq _                                      = empty
 
-let' :: (Carrier sig m, Member Core sig) => User -> m a
-let' = send . Let
+unseqs :: Member Core sig => Term sig a -> NonEmpty (Term sig a)
+unseqs = go
+  where go t = case unseq t of
+          Just (l, r) -> go l <> go r
+          Nothing     -> t :| []
 
-block :: (Foldable t, Carrier sig m, Member Core sig) => t (m a) -> m a
-block = maybe unit getBlock . foldMap (Just . Block)
+(>>>=) :: (Eq a, Carrier sig m, Member Core sig) => (Named a :<- m a) -> m a -> m a
+Named u n :<- a >>>= b = send (Named u a :>>= abstract1 n b)
 
-newtype Block m a = Block { getBlock :: m a }
+infixr 1 >>>=
 
-instance (Carrier sig m, Member Core sig) => Semigroup (Block m a) where
-  Block a <> Block b = Block (send (a :>> b))
+unbind :: (Alternative m, Member Core sig, RightModule sig) => a -> Term sig a -> m (Named a :<- Term sig a, Term sig a)
+unbind n (Term sig) | Just (Named u a :>>= b) <- prj sig = pure (Named u n :<- a, instantiate1 (pure n) b)
+unbind _ _                                               = empty
 
-lam :: (Eq a, Carrier sig m, Member Core sig) => Named a -> m a -> m a
-lam (Named u n) b = send (Lam u (bind1 n b))
+unstatement :: (Alternative m, Member Core sig, RightModule sig) => a -> Term sig a -> m (Maybe (Named a) :<- Term sig a, Term sig a)
+unstatement n t = first (first Just) <$> unbind n t <|> first (Nothing :<-) <$> unseq t
 
-lam' :: (Carrier sig m, Member Core sig) => User -> m User -> m User
-lam' u = lam (named' u)
+do' :: (Eq a, Foldable t, Carrier sig m, Member Core sig) => t (Maybe (Named a) :<- m a) -> m a
```

Comment:
This (and its approximate inverse, `unstatements`) allows us to sequence a list of binding and non-binding statements.

Reply:
Nitpick: I’m not a huge fan of a function that looks so close to a keyword… maybe `begin` here, a la Scheme’s precedent?

Reply:
I was intending to evoke the keyword because of the analogy: `do'` : `do` :: `>>>=` : `>>=`. No?